salt/modules/cipher.py

```diff
@@ -0,0 +1,531 @@
+
+Also every call can override the above defaults:
+If a `key` is defined it is used over a `keyfile`.
+
+
+    salt-call cipher.enc 'mypassonlymastercandec' keyfile='/root/.saltcipher'
+'''
+import base64
+import hashlib
+import os
+import salt.utils
+
+REQ_ERROR = {
+    'crypto': None,
+    'crypto_rsa': None,
+    'openssl': None,
+    'libnacl': None,
+}
+
+__virtualname__ = 'cipher'
+
+try:
+    import libnacl.secret
+except ImportError as e:
+    REQ_ERROR['libnacl'] = 'libnacl import error, perhaps missing python libnacl package'
+
+try:
+    from Crypto.Cipher import AES
+    from Crypto.Cipher import XOR
+except ImportError as e:
+    try:
+        # windows and mac might have lowercase crypto ?
+        from crypto.cipher import AES
+    except ImportError as e:
+        REQ_ERROR['crypto'] = 'import error, missing python Crypto package (install pycrypto)'
+
+try:
+    from Crypto.PublicKey import RSA
+    if 'importKey' not in dir(RSA):
+        REQ_ERROR['crypto_rsa'] = 'python lib pycrypto is missing requirements. Likly need new version.'
+except ImportError as e:
+    REQ_ERROR['crypto_rsa'] = 'import error, missing python Crypto package'
+
+
+if salt.utils.is_windows() or salt.utils.which('openssl') is None:
+    REQ_ERROR['openssl'] = 'missing binary openssl'
+
+
+def _get_config(**kwargs):
+    '''
+    Return configuration
+    '''
+    config = {
+        'key': None,
+        'keyfile': '/etc/salt/pki/master/master.pem',
+    }
+    if '__salt__' in globals():
+        config_key = '{0}.config'.format(__virtualname__)
+        config.update(__salt__['config.get'](config_key, {}))
+    for k in set(config.keys()) & set(kwargs.keys()):
+        config[k] = kwargs[k]
+    return config
+
+
+def _get_key(size_limit=None, trim_newline=False, **kwargs):
+    config = _get_config(**kwargs)
+    key = config['key']
+    if not key:
+        with open(config['keyfile'], 'rb') as keyf:
+            key = keyf.read()
+            if trim_newline:
+                key = key.rstrip('\n')
+    key = str(key)
+    #key = key.encode('ascii','ignore')
+    if size_limit and len(key) > size_limit:
+        #m = base64.b64encode(hashlib.sha256(key).digest())
+        m = base64.b64encode(hashlib.md5(key).digest())
+        key = m[:size_limit]
+    return key
+
+
+def enc(data, **kwargs):
+    '''
+    alias for aes_enc
+
+    CLI Examples:
+
+    .. code-block:: bash
+
+        salt-call cipher.enc
+    '''
+    return aes_enc(data=data, **kwargs)
+
+
+def dec(data, **kwargs):
+    '''
+    alias for aes_dec
+
+    CLI Examples:
+
+    .. code-block:: bash
+
+        salt-call cipher.dec
+    '''
+    return aes_dec(data=data, **kwargs)
+
+
+def libnacl_keygen():
+    '''
+    Use libnacl to generate a private key
+
+    CLI Examples:
+
+    .. code-block:: bash
+
+        salt-call --out=newline_values_only cipher.libnacl_keygen > /root/.naclkey
+    '''
+    if REQ_ERROR['libnacl']:
+        raise Exception(REQ_ERROR['libnacl'])
```

Comment:
Can't we be more specific on the exception raised here? `Exception` is too broad.... `salt.exceptions.CommandExecutionError`?